File: app/src/main/java/net/ducksmanager/inducks/coa/CountryListing.java
package net.ducksmanager.inducks.coa;

import android.app.Activity;

import com.koushikdutta.async.future.FutureCallback;

import net.ducksmanager.whattheduck.WhatTheDuck;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Iterator;

public class CountryListing extends CoaListing {

    private static HashMap<String,String> countryNames= new HashMap<>();
    public static boolean hasFullList = false;

    public CountryListing(Activity activity, FutureCallback callback) {
        super(activity, ListType.COUNTRY_LIST, callback);
    }

    public static String getCountryFullName (String shortCountryName) {
        return countryNames.get(shortCountryName);
    }

    private static void resetCountries() {
        countryNames = new HashMap<>();
    }

    private static void addCountry(String shortName, String fullName) {
        countryNames.put(shortName, fullName);
    }

    @Override
    protected String getUrlSuffix() {
        return "/coa/list/countries/{locale}";
    }

    @Override
    protected void processData(String response) throws JSONException {
        if (response != null) {
            JSONObject responseObject = new JSONObject(response);
            resetCountries();
            addCountriesFullList(responseObject);
        }
    }

    @SuppressWarnings("unchecked")
    public static void addCountries(JSONObject countryNames) throws JSONException {
        if (!hasFullList) {
            if (countryNames.has("static")) { // Legacy JSON structure
                countryNames = countryNames.getJSONObject("static").getJSONObject("pays");
            }

            Iterator<String> countryIterator = countryNames.keys();
            while (countryIterator.hasNext()) {
                String shortName = countryIterator.next();
                String fullName = countryNames.getString(shortName);

                addCountry(shortName, fullName);
                if (!shortName.equals("zz") && !WhatTheDuck.coaCollection.hasCountry(shortName)) {
                    WhatTheDuck.coaCollection.addCountry(shortName);
                }
            }
        }
    }

    private static void addCountriesFullList(JSONObject object) throws JSONException {
        addCountries(object);
        hasFullList = true;
    }
}

File: app/src/main/java/net/ducksmanager/inducks/coa/CountryListing.kt
package net.ducksmanager.inducks.coa

import android.app.Activity

import com.koushikdutta.async.future.FutureCallback

import net.ducksmanager.whattheduck.WhatTheDuck

import org.json.JSONException
import org.json.JSONObject

import java.util.HashMap

class CountryListing(activity: Activity, callback: FutureCallback<*>) : CoaListing(activity, CoaListing.ListType.COUNTRY_LIST, callback) {

    protected override val urlSuffix: String
        get() = "/coa/list/countries/{locale}"

    @Throws(JSONException::class)
    override fun processData(response: String?) {
        if (response != null) {
            resetCountries()
            addCountriesFullList(JSONObject(response))
        }
    }

    companion object {

        private var countryNames = HashMap<String, String>()
        var hasFullList = false

        fun getCountryFullName(shortCountryName: String): String {
            return countryNames[shortCountryName]
        }

        private fun resetCountries() {
            countryNames = HashMap()
        }

        private fun addCountry(shortName: String, fullName: String) {
            countryNames[shortName] = fullName
        }

        @Throws(JSONException::class)
        fun addCountries(countryNames: JSONObject) {
            var countryNames = countryNames
            if (!hasFullList) {
                if (countryNames.has("static")) { // Legacy JSON structure
                    countryNames = countryNames.getJSONObject("static").getJSONObject("pays")
                }

                val countryIterator = countryNames.keys()
                while (countryIterator.hasNext()) {
                    val shortName = countryIterator.next()
                    val fullName = countryNames.getString(shortName)

                    addCountry(shortName, fullName)
                    if (shortName != "zz" && !WhatTheDuck.coaCollection.hasCountry(shortName)) {
                        WhatTheDuck.coaCollection.addCountry(shortName)
                    }
                }
            }
        }

        @Throws(JSONException::class)
        private fun addCountriesFullList(`object`: JSONObject) {
            addCountries(`object`)
            hasFullList = true
        }
    }
}
